text/2022-07-25-PROPOSAL-build-glyphs-in-rust.md

```diff
@@ -0,0 +1,128 @@
+This makes font development projects slower and more expensive.
+
+To make font development faster and cheaper, make all font builds at least an order (preferably two) faster
+
+   * A two order speedup would make edit/build/test loops possible where today you take rest of the day off
+   * Bonus points for using less resources; today large builds exhaust github runners
+
+Prior art around moving subsetting from Python to C++ suggests running two orders faster while consuming an order less resources is plausible.
+
+For maximum impact, do it horizontally, impacting as close as possible to all Google Fonts projects without having to touch them one by one to alter their bespoke build processes. If we must touch them, we should move them to a consistent build structure.
+
+Bias to heavy lifting done in Rust as part of Oxidize. A layer of Python orchestration is acceptable.
+
+Bias to incremental progress, no big bang integration.
+Writing a new compiler in isolation while everyone continues to use fontmake, and then abruptly switching projects over
+(big bang integration) delays impact and increases risk.
+An approach that incrementally improves fontmake has significant appeal.
+
+## Proposal
+
+Migrate the core for each glyph loop
+([varLib._add_gvar](https://github.com/fonttools/fonttools/blob/455158f2bfd9eae2135a5f85bb96549a545cac82/Lib/fontTools/varLib/__init__.py#L222))
+in fontmake from Python to Rust.
+
+* Expose Rust to Python.
+* Let Rust handle the parallelism.
+* Build directly from UFO sources, optionally passing through some intermediary representation.
+* Do NOT use the static TTFs at master locations to build glyph tables.
+  * the static TTFs can still exist, but their glyph work should be brought as close as possible to zero (empty glyphs?)
+
+The next step is to get approval from the fontmake owners to make such changes. @anthrotype basically :D
+
+## Context
+
+### How do you make a variable font anyway
+
+Complicating this, the core approach we take to build a variable font is excitingly indirect (https://simoncozens.github.io/compiling-variable-fonts/):
+
+   1. Generate static TTFs at master locations
+   1. Merge those TTFs to create a final VF
+
+As far as the author can tell there is no requirement for the statics to ever exist, we do it this way purely because
+it was a relatively simple way to get fontmake to support VF. In the end we should go from sources → final VF without this step.
+
+A concrete example may help, consider the N in [Josefin Slab](https://fonts.google.com/specimen/Josefin+Slab)
+(https://github.com/TypeNetwork/Josefinslab) upright:
+
+* [sources/JosefinSlab-Thin.ufo/glyphs/N_.glif](https://github.com/davelab6/josefinslab/blob/master/sources/JosefinSlab-Thin.ufo/glyphs/N_.glif) defines the curve at thin (N)
+* [sources/JosefinSlab-Bold.ufo/glyphs/N_.glif](https://github.com/davelab6/josefinslab/blob/master/sources/JosefinSlab-Bold.ufo/glyphs/N_.glif) defines the curve at bold (N)
```

Comment:
Curves plural

Reply:
Updated to outline contours per your earlier comment